### PR TITLE
upgrade Go to 1.20.6

### LINKED ISSUE #35927: Fix coverage report.
Currently the coverage reports are not working.
To run unit tests with coverage report for Filebeat:
```sh
cd filebeat
TEST_COVERAGE=true mage GoUnitTest
```

You'll see the following error:
```
cover: can't read "github.com/elastic/beats/v7/filebeat/input/syslog/rfc3164_parser.rl": open /home/tiago/devel/beats/filebeat/input/syslog/rfc3164_parser.rl: no such file or directory
Error: failed to write HTML code coverage report: running "go tool cover -html=build/TEST-go-unit.cov -o build/TEST-go-unit.html" failed with exit code 1
```

The problem seems to be caused by some files in the syslog input (like `filebeat/input/syslog/rfc3164_parser.go`) that contain annotations to some `.rl` files, those annotations look like this:
```
//line parser/rfc3164_parser.rl:28
```
Those files are autogenerated.

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

 - Upgrade Go to 1.20.6
 - fixes test coverage #35927 by moving `.rl` files on ...

## Why is it important?

 - To keep up with the latest Go improvements, features and bug fixes.
 - Test coverage is broken due to the usage of `//line` directives on generated code which makes `go tool cover` to look for files in the wrong path, failing to generate the coverage report.

## Checklist

- ~~[ ] My code follows the style guidelines of this project~~
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #35927

### CODE REVIEW

**Review by cmacknz — APPROVED** [review]

**Comment by cmacknz** [comment]
I suspect we'll want the changes in https://github.com/elastic/elastic-agent/pull/3053/commits/db1c47f32a252901bdabd2295818bad581890dd8 here as well since the code is the same.

I also suspect the unit tests in the dev-tool directory aren't being run by CI, because those tests are failing with Go 1.20.5 locally on main.

**Comment by cmacknz** [comment]
```
[2023-07-24T16:44:49.500Z] cover: can't read "github.com/elastic/beats/v7/filebeat/input/syslog/format_check.rl": open /var/lib/jenkins/workspace/PR-36000-8-47b742ad-51f3-47c4-8aec-0c151ac4fbea/src/github.com/elastic/beats/filebeat/input/syslog/format_check.rl: no such file or directory

[2023-07-24T16:44:49.500Z] Error: failed to write HTML code coverage report: running "go tool cover -html=build/TEST-go-unit.cov -o build/TEST-go-unit.html" failed with exit code 1

script returned exit code 1
```

It looks like this problem might have existed before this PR but now the error is actually being caught: https://github.com/elastic/beats/issues/35927. This plausibly could have been a bug fix in `go tool cover` as part of supporting code coverage for integration tests.

Not sure how to fix it yet, possibly we need to avoid generating code coverage for the package containing these generated files.

**Comment by jlind23** [comment]
@AndersonQ I saw you ignored the issue in your latest commit, can we still make sure the code coverage is generated?

**Comment by AndersonQ** [comment]
> @AndersonQ I saw you ignored the issue in your latest commit, can we still make sure the code coverage is generated?

I'm only ignoring the error if it's the one described on https://github.com/elastic/beats/issues/35927. Therefore it won't change the behaviour for anything else, only if this specific error happens, that error will be ignored and the CI won't fail

**Comment by cmacknz** [comment]
Go 1.20.6 was just released, you can bump the Go versions here to 1.20.6 to save from having to follow up with another PR doing that.

File: libbeat/reader/syslog/rfc3164_test.go [hunk 1]
```diff
@@ -156,7 +156,7 @@ func TestParseRFC3164(t *testing.T) {
 				hostname: "test-host",
 				msg:      "this is the message",
 			},
-			wantErr: `validation error at position 5: parsing time "24-08-2003T05:14:15-07:00" as "2006-01-02T15:04:05.999999999Z07:00": cannot parse "8-2003T05:14:15-07:00" as "2006"`,
+			wantErr: `validation error at position 5: parsing time "24-08-2003T05:14:15-07:00" as "2006-01-02T15:04:05.999999999Z07:00": cannot parse "24-08-2003T05:14:15-07:00" as "2006"`,
```
Comment (cmacknz) [inline libbeat/reader/syslog/rfc3164_test.go:159]: Why did this change? Are you sure this is a not a functional change in the parser that caused this? Was there a change in Go 1.20 that needed this?

I want to make sure this is not a hint that something has changed in how we parse values that we need to account for before we release this.

**Review by pierrehilbert — APPROVED** [review]

**Pull request merged**